### DISCUSSION TITLE
Getting CORS while using Analytics on Web

### DISCUSSION THREAD

**Discussion by PradyumnaKrishna** [invertase/react-native-firebase#7982] [Q&A]
So, I am trying to build cross platform application and using react-native-firebase to log event through `firebase/analytics`. I tried the demo on android, it worked fine, but on the web it gave me error 

```
No Firebase App '[DEFAULT]' has been created - call firebase.initializeApp()
```

I managed to initialise the Firebase App by providing it credentials. Now when I tried to log events using `react-native-firebase/analytics`, it gives me CORS error during API call with google analytics.

Not sure if there is something wrong with the setup. What is the correct way to setup the firebase for web using `react-native-firebase`?
Take this code as reference for my setup:

```js
if (Platform.OS === "web") {
  const firebaseConfig = {
    clientId: '',
    appId: "",
    apiKey: "",
    databaseURL: '',
    storageBucket: "",
    messagingSenderId: "",
    projectId: "",
  };
  
  // Initialize Firebase
  firebase.setReactNativeAsyncStorage(AsyncStorage);
  firebase.initializeApp(firebaseConfig)

  analytics().logEvent("home_page_view");
}
```

**Comment by ristewar** [comment]
Did you ever figure this out?

**Reply by ristewar** [reply]
I dug into this a bit more. The problem seems to be that the Firebase Analytics implementation for web is triggering a CORS preflight check, but Google Analytics doesn't support CORS so Chrome fails the preflight check with:

`Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
`

The Analytics web implemention  is setting `no-cors` [here](https://github.com/invertase/react-native-firebase/blob/main/packages/analytics/lib/web/api.js#L317) however it's also passing headers that aren't in the CORS safe-listed request header list (see [here](https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_request_header))  and so Chrome is sending a CORS preflight, which is triggering the CORS failure.

I suspect if the Analytics code were to drop some of the non-safe-listed headers then the `no-cors` would be respected by Chrome.

@Salakar since he added this code in #7899, maybe he can clarify whether Analytics is expected to work for Web or not.

**Reply by ristewar** [reply]
I commented out the two 'no-cache' headers and then the Google Analytics request works fine, I get a 204 back from Google.

```
        const response = await fetch(url, {
          method: 'POST',
          headers: {
            accept: '*/*',
            'accept-encoding': 'gzip, deflate, br',
            'Content-Type': 'text/plain;charset=UTF-8',
            'accept-language': 'en-US,en;q=0.9',
//            'cache-control': 'no-cache',
            'content-length': '0',
            origin: 'firebase',
//            pragma: 'no-cache',
            'sec-fetch-dest': 'empty',
            'sec-fetch-mode': 'no-cors',
            'sec-fetch-site': 'cross-site',
            'user-agent': 'react-native-firebase',
          },
        });
```

So apparently either of those `no-cache` headers will cause Chrome to send a CORS preflight even when `no-cors` is set.

**Reply by mikehardy** [reply]
Huh - analytics is definitely supposed to work on the web. Let's take that as a given.

Are we certain google analytics doesn't support CORS if we turn off no-cors? (perhaps by also tweaking the headers slightly?)


Excellent work finding a set of things in the request that *does* work  with no-cors set, however this may have problems with intermediate proxies perhaps 🤔 

What are the headers you see come back from the request? If they make it appear that the document is always expired etc (or also has no-cache headers) then I would have less fear of caching issues despite not sending the cache control stuff in the request

**Reply by ristewar** [reply]
Welp, it's apparently because `fetch()` requires us to set `mode: 'no-cors'`. Just setting it in `headers` isn't sufficient. Adding that one line fixes the issue.

```
const response = await fetch(url, {
      method: 'POST',
+      mode: 'no-cors',
      headers: {
        accept: '*/*',
        'accept-encoding': 'gzip, deflate, br',
        'Content-Type': 'text/plain;charset=UTF-8',
        'accept-language': 'en-US,en;q=0.9',
        'cache-control': 'no-cache',
        'content-length': '0',
        origin: 'firebase',
        pragma: 'no-cache',
        'sec-fetch-dest': 'empty',
        'sec-fetch-mode': 'no-cors',
        'sec-fetch-site': 'cross-site',
        'user-agent': 'react-native-firebase',
      },
    });
```

**Reply by mikehardy** [reply]
ah ha! if you post a PR and do the license-bot-signing-thingie I could merge and release that in about 1 minute
I'm aware of that setting, forgot it was a setting vs a header tho

**Reply by ristewar** [reply]
Sent PR https://github.com/invertase/react-native-firebase/pull/8238
